src/main/kotlin/org/rust/ide/inspections/RsSelfConventionInspection.kt

```diff
@@ -13,49 +13,53 @@ import org.rust.lang.core.psi.RsTraitItem
+import org.rust.stdext.typeAscription
 
 class RsSelfConventionInspection : RsLocalInspectionTool() {
 
     override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean) =
         object : RsVisitor() {
             override fun visitFunction(m: RsFunction) {
                 val owner = m.owner
-                if (!(owner.isInherentImpl || owner.isTrait)) return
+                val traitOrImpl = when (owner) {
+                    is RsFunctionOwner.Trait -> typeAscription<RsTraitOrImpl>(owner.trait)
+                    is RsFunctionOwner.Impl -> owner.impl.takeIf { owner.isInherent }
+                    else -> null
+                } ?: return
 
                 val convention = SELF_CONVENTIONS.find { m.identifier.text.startsWith(it.prefix) } ?: return
-                if (m.selfType in convention.selfTypes) return
-                if (m.selfType == SelfType.SELF && owner.isOwnerCopyable()) return
+                if (m.selfSignature in convention.selfSignatures) return
+
+                if (m.selfSignature == SelfSignature.BY_VAL) {
+                    val selfType = traitOrImpl.selfType
+                    val implLookup = ImplLookup.relativeTo(traitOrImpl)
+                    if (selfType is TyUnknown || implLookup.isCopy(selfType)) return
+                }
+
                 holder.registerProblem(m.selfParameter ?: m.identifier, convention)
             }
         }
 
-    private fun RsFunctionOwner.isOwnerCopyable(): Boolean {
-        val impls = when (this) {
-            is RsFunctionOwner.Trait -> trait.traits
-            is RsFunctionOwner.Impl -> impl.traits
-            else -> null
-        } ?: return false
-        return impls.any { it.name == "Copy" }
-    }
-
     private companion object {
         val SELF_CONVENTIONS = listOf(
-            SelfConvention("as_", listOf(SelfType.REF_SELF, SelfType.REF_MUT_SELF)),
-            SelfConvention("from_", listOf(SelfType.NO_SELF)),
-            SelfConvention("into_", listOf(SelfType.SELF)),
-            SelfConvention("is_", listOf(SelfType.REF_SELF, SelfType.NO_SELF)),
-            SelfConvention("to_", listOf(SelfType.REF_SELF))
+            SelfConvention("as_", listOf(SelfSignature.BY_REF, SelfSignature.BY_MUT_REF)),
+            SelfConvention("from_", listOf(SelfSignature.NO_SELF)),
+            SelfConvention("into_", listOf(SelfSignature.BY_VAL)),
+            SelfConvention("is_", listOf(SelfSignature.BY_REF, SelfSignature.NO_SELF)),
+            SelfConvention("to_", listOf(SelfSignature.BY_REF))
         )
     }
 }
 
-enum class SelfType(val description: String) {
+enum class SelfSignature(val description: String) {
     NO_SELF("no self"),
-    SELF("self by value"),
-    REF_SELF("self by reference"),
-    REF_MUT_SELF("self by mutable reference");
+    BY_VAL("self by value"),
+    BY_REF("self by reference"),
+    BY_MUT_REF("self by mutable reference");
 }
 
 private val RsTraitOrImpl.traits: Sequence<RsTraitItem>
```

Comment:
You can remove this function and make supertraits private again.

Reply:
Excellent, thanks @farodin91 !